File: fsharp/Sort.fs
namespace Benchmarks

open System

[<AbstractClass>]
type SortBenchmark() =
    inherit Benchmark()

    let mutable data = Array.empty<int>
    let mutable size = 0L
    let mutable result = 0u

    member _.Data with get() = data and set(v) = data <- v
    member _.Size with get() = size and set(v) = size <- v
    member _.Result with get() = result and set(v) = result <- v

    override this.Prepare() =
        size <- this.ConfigVal("size")
        data <- Array.init (int size) (fun _ -> Helper.NextInt(1000000))
        result <- 0u

    override this.Run(_: int64) =
        result <- result + uint32 data.[Helper.NextInt(int size)]
        let t = this.Test()
        result <- result + uint32 t.[Helper.NextInt(int size)]

    override this.Checksum = result

    abstract member Test: unit -> int[]

type SortMerge() =
    inherit SortBenchmark()

    let mergeSortInplace (arr: int[]) =
        let temp = Array.zeroCreate<int> arr.Length

        let rec mergeSortHelper left right =
            if left >= right then ()
            else
                let mid = (left + right) / 2
                mergeSortHelper left mid
                mergeSortHelper (mid + 1) right
                merge arr temp left mid right

        and merge (arr: int[]) (temp: int[]) left mid right =

            for i = left to right do
                temp.[i] <- arr.[i]

            let mutable iIdx = left
            let mutable jIdx = mid + 1
            let mutable k = left

            while iIdx <= mid && jIdx <= right do
                if temp.[iIdx] <= temp.[jIdx] then
                    arr.[k] <- temp.[iIdx]
                    iIdx <- iIdx + 1
                else
                    arr.[k] <- temp.[jIdx]
                    jIdx <- jIdx + 1
                k <- k + 1

            while iIdx <= mid do
                arr.[k] <- temp.[iIdx]
                iIdx <- iIdx + 1
                k <- k + 1

        mergeSortHelper 0 (arr.Length - 1)

    override this.Test() =
        let arr = Array.copy this.Data
        mergeSortInplace arr
        arr

type SortQuick() =
    inherit SortBenchmark()

    let quickSort (arr: int[]) =
        let rec sort low high =
            if low >= high then ()
            else
                let pivot = arr.[(low + high) / 2]
                let mutable i = low
                let mutable j = high

                while i <= j do
                    while arr.[i] < pivot do
                        i <- i + 1

                    while arr.[j] > pivot do
                        j <- j - 1

                    if i <= j then
                        let temp = arr.[i]
                        arr.[i] <- arr.[j]
                        arr.[j] <- temp
                        i <- i + 1
                        j <- j - 1

                sort low j
                sort i high

        sort 0 (arr.Length - 1)

    override this.Test() =
        let arr = Array.copy this.Data
        quickSort arr
        arr

type SortSelf() =
    inherit SortBenchmark()

    override this.Test() =
        let arr = Array.copy this.Data
        Array.Sort(arr)
        arr

File: julia/src/benchmarks/Sort.jl
using ..BenchmarkFramework

abstract type AbstractSortBenchmark <: AbstractBenchmark end

mutable struct SortBenchmark <: AbstractSortBenchmark
    size::Int64
    data::Vector{Int32}
    result::UInt32

    function SortBenchmark()
        size_val = Helper.config_i64("SortBenchmark", "size")
        new(size_val, Int32[], UInt32(0))
    end
end

name(b::SortBenchmark)::String = "SortBenchmark"

function prepare(b::SortBenchmark)
    empty!(b.data)
    for _ in 1:b.size
        push!(b.data, Helper.next_int(1_000_000))
    end
end

function run(b::SortBenchmark, iteration_id::Int64)

    error("Abstract method 'test' not implemented")
end

function test(b::SortBenchmark)::Vector{Int32}

    error("Abstract method 'test' not implemented")
end

function checksum(b::SortBenchmark)::UInt32
    return b.result
end

mutable struct SortQuick <: AbstractSortBenchmark
    size::Int64
    data::Vector{Int32}
    result::UInt32

    function SortQuick()
        size_val = Helper.config_i64("SortQuick", "size")
        new(size_val, Int32[], UInt32(0))
    end
end

name(b::SortQuick)::String = "SortQuick"

function prepare(b::SortQuick)
    empty!(b.data)
    for _ in 1:b.size
        push!(b.data, Helper.next_int(1_000_000))
    end
end

function quick_sort!(arr::Vector{Int32}, low::Int, high::Int)
    if low >= high
        return
    end

    pivot = arr[(low + high) ÷ 2]
    i = low
    j = high

    while i <= j
        while arr[i] < pivot
            i += 1
        end
        while arr[j] > pivot
            j -= 1
        end
        if i <= j
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1
        end
    end

    quick_sort!(arr, low, j)
    quick_sort!(arr, i, high)
end

function test(b::SortQuick)::Vector{Int32}
    arr = copy(b.data)
    if length(arr) > 0
        quick_sort!(arr, 1, length(arr))  
    end
    return arr
end

function run(b::SortQuick, iteration_id::Int64)

    if length(b.data) > 0
        idx = Helper.next_int(length(b.data)) + 1  
        b.result = (b.result + UInt32(b.data[idx])) & 0xffffffff
    end

    sorted_arr = test(b)
    if length(sorted_arr) > 0
        idx = Helper.next_int(length(sorted_arr)) + 1  
        b.result = (b.result + UInt32(sorted_arr[idx])) & 0xffffffff
    end
end

mutable struct SortMerge <: AbstractSortBenchmark
    size::Int64
    data::Vector{Int32}
    result::UInt32

    function SortMerge()
        size_val = Helper.config_i64("SortMerge", "size")
        new(size_val, Int32[], UInt32(0))
    end
end

name(b::SortMerge)::String = "SortMerge"

function prepare(b::SortMerge)
    empty!(b.data)
    for _ in 1:b.size
        push!(b.data, Helper.next_int(1_000_000))
    end
end

function merge_sort!(arr::Vector{Int32})
    if length(arr) <= 1
        return arr
    end

    temp = similar(arr)
    merge_sort_helper!(arr, temp, 1, length(arr))  
    return arr
end

function merge_sort_helper!(arr::Vector{Int32}, temp::Vector{Int32}, left::Int, right::Int)
    if left >= right
        return
    end

    mid = (left + right) ÷ 2
    merge_sort_helper!(arr, temp, left, mid)
    merge_sort_helper!(arr, temp, mid + 1, right)
    merge!(arr, temp, left, mid, right)
end

function merge!(arr::Vector{Int32}, temp::Vector{Int32}, left::Int, mid::Int, right::Int)

    for i in left:right
        temp[i] = arr[i]
    end

    i = left      
    j = mid + 1   
    k = left      

    while i <= mid && j <= right
        if temp[i] <= temp[j]
            arr[k] = temp[i]
            i += 1
        else
            arr[k] = temp[j]
            j += 1
        end
        k += 1
    end

    while i <= mid
        arr[k] = temp[i]
        i += 1
        k += 1
    end
end

function test(b::SortMerge)::Vector{Int32}
    arr = copy(b.data)
    if length(arr) > 0
        merge_sort!(arr)
    end
    return arr
end

function run(b::SortMerge, iteration_id::Int64)

    if length(b.data) > 0
        idx = Helper.next_int(length(b.data)) + 1  
        b.result = (b.result + UInt32(b.data[idx])) & 0xffffffff
    end

    sorted_arr = test(b)
    if length(sorted_arr) > 0
        idx = Helper.next_int(length(sorted_arr)) + 1  
        b.result = (b.result + UInt32(sorted_arr[idx])) & 0xffffffff
    end
end

mutable struct SortSelf <: AbstractSortBenchmark
    size::Int64
    data::Vector{Int32}
    result::UInt32

    function SortSelf()
        size_val = Helper.config_i64("SortSelf", "size")
        new(size_val, Int32[], UInt32(0))
    end
end

name(b::SortSelf)::String = "SortSelf"

function prepare(b::SortSelf)
    empty!(b.data)
    for _ in 1:b.size
        push!(b.data, Helper.next_int(1_000_000))
    end
end

function test(b::SortSelf)::Vector{Int32}
    arr = copy(b.data)
    sort!(arr)
    return arr
end

function run(b::SortSelf, iteration_id::Int64)

    if length(b.data) > 0
        idx = Helper.next_int(length(b.data)) + 1  
        b.result = (b.result + UInt32(b.data[idx])) & 0xffffffff
    end

    sorted_arr = test(b)
    if length(sorted_arr) > 0
        idx = Helper.next_int(length(sorted_arr)) + 1  
        b.result = (b.result + UInt32(sorted_arr[idx])) & 0xffffffff
    end
end

function checksum(b::AbstractSortBenchmark)::UInt32
    return b.result
end